File: apps/cli/internal/workflow/parser.go
package workflow

import (
	"bytes"
	"fmt"
	"os"
	"path/filepath"
	"sort"
	"strings"

	"github.com/goccy/go-yaml"
)

const (
	// maxWorkflowSizeBytes is the maximum allowed size for a workflow file (1MB)
	// This prevents resource exhaustion from maliciously large files
	maxWorkflowSizeBytes = 1 * 1024 * 1024
)

// validateWorkflowContent checks for potentially malicious or malformed content.
// This provides defense-in-depth against crafted workflow files.
func validateWorkflowContent(data []byte) error {
	// Size limit to prevent resource exhaustion
	if len(data) > maxWorkflowSizeBytes {
		return fmt.Errorf("workflow file exceeds maximum size of %d bytes", maxWorkflowSizeBytes)
	}

	// Null bytes indicate binary content disguised as YAML
	if bytes.Contains(data, []byte{0x00}) {
		return fmt.Errorf("workflow file contains null bytes (binary content not allowed)")
	}

	// Check for excessive control characters (excluding newline, carriage return, tab)
	// This catches malformed files that might exploit YAML parser edge cases
	controlCount := 0
	for _, b := range data {
		if b < 32 && b != '\n' && b != '\r' && b != '\t' {
			controlCount++
		}
	}
	if controlCount > 10 {
		return fmt.Errorf("workflow file contains excessive control characters (%d found)", controlCount)
	}

	return nil
}

// ParseWorkflowFile reads and parses a single workflow YAML file.
// The path must be validated by the caller to be within an expected directory.
func ParseWorkflowFile(path string) (*Workflow, error) {
	data, err := os.ReadFile(path) //nolint:gosec // path validated by caller via DiscoverWorkflows
	if err != nil {
		return nil, fmt.Errorf("reading workflow file: %w", err)
	}

	// Validate content before parsing (defense-in-depth)
	if err := validateWorkflowContent(data); err != nil {
		return nil, err
	}

	var wf Workflow
	if err := yaml.Unmarshal(data, &wf); err != nil {
		return nil, fmt.Errorf("parsing workflow YAML: %w", err)
	}

	return &wf, nil
}

// DiscoverWorkflows finds all workflow files in a directory.
// Only regular files with .yml or .yaml extensions are returned.
// Symlinks and files outside the directory are skipped for security.
func DiscoverWorkflows(dir string) ([]string, error) {
	if dir == "" {
		return nil, fmt.Errorf("workflows directory cannot be empty")
	}

	absDir, err := filepath.Abs(dir)
	if err != nil {
		return nil, fmt.Errorf("resolving workflows directory: %w", err)
	}

	entries, err := os.ReadDir(dir)
	if err != nil {
		return nil, fmt.Errorf("reading workflows directory: %w", err)
	}

	var workflows []string
	for _, entry := range entries {
		if entry.IsDir() {
			continue
		}

		// Skip symlinks to prevent path traversal
		if entry.Type()&os.ModeSymlink != 0 {
			continue
		}

		ext := filepath.Ext(entry.Name())
		if ext != ".yml" && ext != ".yaml" {
			continue
		}

		fullPath := filepath.Join(dir, entry.Name())
		absPath, err := filepath.Abs(fullPath)
		if err != nil {
			continue
		}

		// Ensure resolved path is within the directory using filepath.Rel
		relPath, err := filepath.Rel(absDir, absPath)
		if err != nil || strings.HasPrefix(relPath, "..") {
			continue
		}

		workflows = append(workflows, fullPath)
	}

	return workflows, nil
}

// ExtractJobInfo extracts job information from a workflow for TUI display.
// Returns a slice of JobInfo with ID and display name for each job.
// Jobs are sorted topologically by needs dependencies, with alphabetical ordering
// as a tiebreaker for jobs at the same dependency level.
func ExtractJobInfo(wf *Workflow) []JobInfo {
	if wf == nil || wf.Jobs == nil {
		return nil
	}

	// First pass: build job info map and parse needs
	jobInfoMap := make(map[string]JobInfo)
	for id, job := range wf.Jobs {
		if job == nil {
			continue
		}

		name := job.Name
		if name == "" {
			name = id
		}

		// Parse needs (can be string or []string)
		var needs []string
		switch n := job.Needs.(type) {
		case string:
			if n != "" {
				needs = []string{n}
			}
		case []any:
			for _, v := range n {
				if s, ok := v.(string); ok {
					needs = append(needs, s)
				}
			}
		}

		jobInfoMap[id] = JobInfo{
			ID:    id,
			Name:  name,
			Needs: needs,
		}
	}

	// Topological sort using Kahn's algorithm
	return topologicalSort(jobInfoMap)
}

// topologicalSort performs a topological sort of jobs based on their needs dependencies.
// Jobs with no dependencies come first, followed by jobs that depend on them.
// Within each level, jobs are sorted alphabetically for deterministic ordering.
func topologicalSort(jobInfoMap map[string]JobInfo) []JobInfo {
	if len(jobInfoMap) == 0 {
		return nil
	}

	// Calculate in-degree (number of dependencies) for each job
	inDegree := make(map[string]int)
	for id := range jobInfoMap {
		inDegree[id] = 0
	}

	// Build adjacency list and count in-degrees
	// dependents[a] contains all jobs that depend on a
	dependents := make(map[string][]string)
	for id, job := range jobInfoMap {
		for _, dep := range job.Needs {
			// Only count dependencies that exist in the workflow
			if _, exists := jobInfoMap[dep]; exists {
				inDegree[id]++
				dependents[dep] = append(dependents[dep], id)
			}
		}
	}

	// Start with jobs that have no dependencies
	var queue []string
	for id, degree := range inDegree {
		if degree == 0 {
			queue = append(queue, id)
		}
	}
	// Sort for deterministic ordering
	sort.Strings(queue)

	var result []JobInfo
	for len(queue) > 0 {
		// Process all jobs at current level
		current := queue[0]
		queue = queue[1:]

		result = append(result, jobInfoMap[current])

		// Find all dependents and reduce their in-degree
		var nextLevel []string
		for _, dependent := range dependents[current] {
			inDegree[dependent]--
			if inDegree[dependent] == 0 {
				nextLevel = append(nextLevel, dependent)
			}
		}

		// Sort next level for deterministic ordering and add to queue
		sort.Strings(nextLevel)
		queue = append(queue, nextLevel...)
	}

	// Handle cycles or missing dependencies - add remaining jobs alphabetically
	if len(result) < len(jobInfoMap) {
		var remaining []string
		addedSet := make(map[string]bool)
		for _, job := range result {
			addedSet[job.ID] = true
		}
		for id := range jobInfoMap {
			if !addedSet[id] {
				remaining = append(remaining, id)
			}
		}
		sort.Strings(remaining)
		for _, id := range remaining {
			result = append(result, jobInfoMap[id])
		}
	}

	return result
}

// ExtractJobInfoFromDir discovers and parses all workflows in a directory,
// returning job info for all jobs across all workflows.
func ExtractJobInfoFromDir(dir string) ([]JobInfo, error) {
	workflows, err := DiscoverWorkflows(dir)
	if err != nil {
		return nil, err
	}

	var allJobs []JobInfo
	for _, wfPath := range workflows {
		wf, err := ParseWorkflowFile(wfPath)
		if err != nil {
			continue // Skip workflows that fail to parse
		}
		jobs := ExtractJobInfo(wf)
		allJobs = append(allJobs, jobs...)
	}

	return allJobs, nil
}

File: apps/cli/src/commands/mock/workflow/parser.ts
import { readdir, readFile } from "node:fs/promises";
import { join } from "node:path";
import { load } from "js-yaml";
import type { Job, JobInfo, Workflow } from "./types.js";

/**
 * Maximum allowed size for a workflow file (1MB).
 * Prevents resource exhaustion from maliciously large files.
 */
const MAX_WORKFLOW_SIZE_BYTES = 1 * 1024 * 1024;

/**
 * Validates workflow content for potentially malicious or malformed content.
 */
const validateWorkflowContent = (data: string): void => {
  if (data.length > MAX_WORKFLOW_SIZE_BYTES) {
    throw new Error(
      `Workflow file exceeds maximum size of ${MAX_WORKFLOW_SIZE_BYTES} bytes`
    );
  }

  if (data.includes("\0")) {
    throw new Error(
      "Workflow file contains null bytes (binary content not allowed)"
    );
  }

  let controlCount = 0;
  for (const char of data) {
    const code = char.charCodeAt(0);
    if (code < 32 && code !== 10 && code !== 13 && code !== 9) {
      controlCount++;
    }
  }
  if (controlCount > 10) {
    throw new Error(
      `Workflow file contains excessive control characters (${controlCount} found)`
    );
  }
};

/**
 * Parses a workflow YAML string into a Workflow object.
 *
 * @param content - Raw YAML content
 * @returns Parsed workflow
 */
export const parseWorkflow = (content: string): Workflow => {
  validateWorkflowContent(content);

  const parsed = load(content);

  if (!parsed || typeof parsed !== "object" || Array.isArray(parsed)) {
    throw new Error("Workflow YAML must be an object");
  }

  const workflow = parsed as Record<string, unknown>;

  if (!workflow.jobs || typeof workflow.jobs !== "object") {
    throw new Error("Workflow must have a 'jobs' object");
  }

  return workflow as unknown as Workflow;
};

/**
 * Parses needs field which can be a string or array.
 */
const parseNeeds = (needs: unknown): readonly string[] => {
  if (!needs) {
    return [];
  }

  if (typeof needs === "string") {
    return needs ? [needs] : [];
  }

  if (Array.isArray(needs)) {
    return needs.filter((n): n is string => typeof n === "string");
  }

  return [];
};

interface DependencyGraph {
  inDegree: Map<string, number>;
  dependents: Map<string, string[]>;
}

/**
 * Builds the dependency graph for topological sorting.
 * Returns in-degree map and adjacency list of dependents.
 */
const buildDependencyGraph = (
  jobInfoMap: Map<string, JobInfo>
): DependencyGraph => {
  const inDegree = new Map<string, number>();
  const dependents = new Map<string, string[]>();

  for (const id of jobInfoMap.keys()) {
    inDegree.set(id, 0);
  }

  for (const [id, job] of jobInfoMap) {
    for (const dep of job.needs) {
      if (!jobInfoMap.has(dep)) {
        continue;
      }
      inDegree.set(id, (inDegree.get(id) ?? 0) + 1);
      const existing = dependents.get(dep) ?? [];
      existing.push(id);
      dependents.set(dep, existing);
    }
  }

  return { inDegree, dependents };
};

/**
 * Finds all jobs with zero in-degree (no dependencies).
 */
const findRootJobs = (inDegree: Map<string, number>): string[] => {
  const roots: string[] = [];
  for (const [id, degree] of inDegree) {
    if (degree === 0) {
      roots.push(id);
    }
  }
  return roots.sort();
};

/**
 * Processes dependents of a job, reducing their in-degree.
 * Returns jobs that are now ready (in-degree became 0).
 */
const processJobDependents = (
  jobId: string,
  dependents: Map<string, string[]>,
  inDegree: Map<string, number>
): string[] => {
  const readyJobs: string[] = [];
  for (const dependent of dependents.get(jobId) ?? []) {
    const newDegree = (inDegree.get(dependent) ?? 1) - 1;
    inDegree.set(dependent, newDegree);
    if (newDegree === 0) {
      readyJobs.push(dependent);
    }
  }
  return readyJobs.sort();
};

/**
 * Adds remaining jobs (from cycles or missing deps) to the result.
 */
const appendRemainingJobs = (
  result: JobInfo[],
  jobInfoMap: Map<string, JobInfo>
): void => {
  if (result.length >= jobInfoMap.size) {
    return;
  }

  const addedSet = new Set(result.map((j) => j.id));
  const remaining = [...jobInfoMap.keys()]
    .filter((id) => !addedSet.has(id))
    .sort();

  for (const id of remaining) {
    const job = jobInfoMap.get(id);
    if (job) {
      result.push(job);
    }
  }
};

/**
 * Performs topological sort of jobs based on their needs dependencies.
 * Jobs with no dependencies come first, followed by jobs that depend on them.
 * Within each level, jobs are sorted alphabetically for deterministic ordering.
 */
const topologicalSort = (
  jobInfoMap: Map<string, JobInfo>
): readonly JobInfo[] => {
  if (jobInfoMap.size === 0) {
    return [];
  }

  const { inDegree, dependents } = buildDependencyGraph(jobInfoMap);
  const queue = findRootJobs(inDegree);
  const result: JobInfo[] = [];

  while (queue.length > 0) {
    const current = queue.shift();
    if (!current) {
      continue;
    }

    const job = jobInfoMap.get(current);
    if (job) {
      result.push(job);
    }

    const readyJobs = processJobDependents(current, dependents, inDegree);
    queue.push(...readyJobs);
  }

  appendRemainingJobs(result, jobInfoMap);

  return result;
};

/**
 * Extracts job information from a workflow for display.
 * Jobs are sorted topologically by needs dependencies.
 *
 * @param workflow - Parsed workflow
 * @returns Array of JobInfo sorted by dependencies
 */
export const extractJobInfo = (workflow: Workflow): readonly JobInfo[] => {
  if (!workflow.jobs) {
    return [];
  }

  const jobInfoMap = new Map<string, JobInfo>();

  for (const [id, job] of Object.entries(workflow.jobs)) {
    if (!job) {
      continue;
    }

    jobInfoMap.set(id, {
      id,
      name: job.name || id,
      needs: parseNeeds(job.needs),
    });
  }

  return topologicalSort(jobInfoMap);
};

/**
 * Discovers workflow files in a directory.
 *
 * @param dir - Path to .github/workflows directory
 * @returns Array of workflow filenames
 */
export const discoverWorkflows = async (
  dir: string
): Promise<readonly string[]> => {
  try {
    const files = await readdir(dir);
    return files
      .filter((file) => file.endsWith(".yml") || file.endsWith(".yaml"))
      .sort();
  } catch {
    return [];
  }
};

/**
 * Result of parsing a workflow file.
 */
export interface ParsedWorkflow {
  readonly filename: string;
  readonly workflow: Workflow;
  readonly jobs: readonly JobInfo[];
}

/**
 * Parses all workflows in a directory.
 *
 * @param workflowsDir - Path to .github/workflows directory
 * @returns Array of parsed workflows
 */
export const parseWorkflowsFromDir = async (
  workflowsDir: string
): Promise<readonly ParsedWorkflow[]> => {
  const filenames = await discoverWorkflows(workflowsDir);
  const results: ParsedWorkflow[] = [];

  for (const filename of filenames) {
    try {
      const content = await readFile(join(workflowsDir, filename), "utf-8");
      const workflow = parseWorkflow(content);
      const jobs = extractJobInfo(workflow);
      results.push({ filename, workflow, jobs });
    } catch {
      // Skip workflows that fail to parse
    }
  }

  return results;
};

/**
 * Gets a specific job from a workflow.
 *
 * @param workflow - Parsed workflow
 * @param jobId - Job ID to find
 * @returns The job if found
 */
export const getJob = (workflow: Workflow, jobId: string): Job | undefined => {
  return workflow.jobs[jobId];
};
